src/lib.rs

```diff
@@ -48,5 +53,73 @@
 #![cfg_attr(feature="clippy", deny(clippy, unicode_not_nfc, wrong_pub_self_convention,
                                    option_unwrap_used))]
 
-#[test]
-fn place_holder() {}
+
+#[cfg(target_os = "macos")]
+extern crate libc;
+
+mod app;
+pub use app::App;
+
+
+
+#[cfg(target_os = "linux")]
+mod linux;
+#[cfg(target_os = "linux")]
+pub use linux::{install, open};
+
+
+
+#[cfg(target_os = "windows")]
+mod windows;
+#[cfg(target_os = "windows")]
+pub use windows::{install, open};
+
+
+
+#[cfg(target_os = "macos")]
+mod macos;
+#[cfg(target_os = "macos")]
+pub use macos::{install, open};
+
+
+/// Foreign Function call Interface to use this library
+pub mod ffi {
+	#![allow(unsafe_code, unused)]
+
+	use std::ffi::CString;
+	use std::os::raw::c_char;
+	use super::{App, open as rust_open, install as rust_install};
+
+	#[no_mangle]
+	/// open the given URI on this system
+	pub unsafe extern "C" fn open(uri: *mut c_char) -> bool {
+		if let Ok(string_uri) = CString::from_raw(uri).into_string() {
+			rust_open(string_uri)
+		} else {
+			false
+		}
+	}
+
+	#[no_mangle]
+	/// install the given App definition for each scheme URI on the system
+	/// schemes are a comma delimited list of schemes
+	pub unsafe extern "C" fn install(
+			bundle_id: *mut c_char,
```

Comment:
These are all utf8 strings ? If yes they should take `*mut u8, usize` for each String.
Also, in that case instead of passing so many parameters, you could just expose a C-App equivalent of rust-App in this crate which the FFI caller can fill and simply pass,